File: Tests/DoNotTypeCoreTests/FallbackTranscriberTests.swift
import XCTest

@testable import DoNotTypeCore

/// A provider that answers after a delay, or fails after one.
private struct SlowProvider: TranscriptionProvider {
    let name: String
    var delay: Duration
    var text: String
    var failure: (any Error)?

    func transcribe(_ request: TranscriptionRequest) async throws -> TranscriptionResult {
        try await Task.sleep(for: delay)
        try Task.checkCancellation()
        if let failure { throw failure }
        return TranscriptionResult(
            transcript: Transcript(transcript: text, language: "en"),
            usage: TokenUsage(), rawOutput: text)
    }
}

final class FallbackTranscriberTests: XCTestCase {
    private let audio = AudioFile(data: Data("wav".utf8), mimeType: "audio/wav")

    private func service(
        _ name: String, delay: Duration, text: String, failure: (any Error)? = nil
    ) -> TranscriptionService {
        TranscriptionService(
            provider: SlowProvider(name: name, delay: delay, text: text, failure: failure),
            model: "\(name)-model", systemInstruction: "i")
    }

    /// The common case: the primary answers normally, so the hedge never fires and never bills.
    func testAFastPrimaryIsNeverSecondGuessed() async throws {
        let hedger = FallbackTranscriber(
            primary: service("primary", delay: .milliseconds(10), text: "primary"),
            secondary: service("secondary", delay: .milliseconds(10), text: "secondary"),
            hedgeAfter: .seconds(30))

        let outcome = try await hedger.transcribe(audio: audio, context: nil)

        XCTAssertEqual(outcome.result.transcript.transcript, "primary")
        XCTAssertFalse(outcome.attribution.wasFallback)
        XCTAssertEqual(outcome.attribution.provider, "primary")
    }

    /// The case this type exists for: the primary stalls, the hedge fires, the user gets words.
    func testAStalledPrimaryIsOvertakenByTheHedge() async throws {
        let hedger = FallbackTranscriber(
            primary: service("primary", delay: .seconds(30), text: "primary"),
            secondary: service("secondary", delay: .milliseconds(20), text: "secondary"),
            hedgeAfter: .milliseconds(20))

        let outcome = try await hedger.transcribe(audio: audio, context: nil)

        XCTAssertEqual(outcome.result.transcript.transcript, "secondary")
        XCTAssertTrue(outcome.attribution.wasFallback, "the caller has to be able to say so")
        XCTAssertEqual(outcome.attribution.provider, "secondary")
    }

    /// Nothing to wait for: a primary that fails hands over rather than burning the hedge delay.
    ///
    /// The delay is a realistic eight seconds and the elapsed time is asserted, because neither
    /// was true before and the gap that hid was a real one. With a 20 ms hedge and only the
    /// transcript checked, this test passed against an implementation that always slept the full
    /// delay: "secondary" comes back either way, just six seconds later. Which backend answered
    /// is not the claim being made here — *when* it started is.
    ///
    /// The failure is a 400 rather than a 500 so that `isTransient` is false and no retry backoff
    /// lands inside the measurement. That is also the shape of the failure this was written for:
    /// Gemini answering `HTTP 400: This API is not available in your current location` in about a
    /// second, after which there is nothing left to wait for.
    func testAFailingPrimaryFallsBackWithoutWaitingOutTheDelay() async throws {
        let hedger = FallbackTranscriber(
            primary: service(
                "primary", delay: .milliseconds(5), text: "",
                failure: ProviderError.http(status: 400, body: "location not supported")),
            secondary: service("secondary", delay: .milliseconds(10), text: "secondary"),
            hedgeAfter: .seconds(8))

        let clock = ContinuousClock()
        let started = clock.now
        let outcome = try await hedger.transcribe(audio: audio, context: nil)
        let elapsed = clock.now - started

        XCTAssertEqual(outcome.result.transcript.transcript, "secondary")
        XCTAssertTrue(outcome.attribution.wasFallback)
        XCTAssertLessThan(
            elapsed, .seconds(1),
            "the hedge waited out its delay after the primary had already failed")
    }

    /// The primary's error is the one that explains the user's configuration, so it is the one
    /// they see when everything fails.
    func testWhenBothFailThePrimaryErrorSurfaces() async {
        let hedger = FallbackTranscriber(
            primary: service(
                "primary", delay: .milliseconds(5), text: "",
                failure: ProviderError.missingAPIKey(envVar: "PRIMARY_KEY")),
            secondary: service(
                "secondary", delay: .milliseconds(5), text: "",
                failure: ProviderError.http(status: 500, body: "boom")),
            hedgeAfter: .milliseconds(1))

        do {
            _ = try await hedger.transcribe(audio: audio, context: nil)
            XCTFail("expected a failure")
        } catch ProviderError.missingAPIKey(let envVar) {
            XCTAssertEqual(envVar, "PRIMARY_KEY")
        } catch {
            XCTFail("expected the primary's error, got \(error)")
        }
    }

    /// No secondary configured is the default, and must behave exactly as before this type existed.
    func testWithoutASecondaryItIsATransparentPassThrough() async throws {
        let hedger = FallbackTranscriber(
            primary: service("primary", delay: .milliseconds(5), text: "primary"))

        let outcome = try await hedger.transcribe(audio: audio, context: nil)
        XCTAssertEqual(outcome.result.transcript.transcript, "primary")
        XCTAssertFalse(outcome.attribution.wasFallback)
    }
}

File: windows/DoNotType.Core.Tests/FallbackTranscriberTests.cs
using System.Diagnostics;
using DoNotType.Core;
using Xunit;

namespace DoNotType.Core.Tests;

/// <summary>
/// The C# port of <c>FallbackTranscriber</c> has to behave like the Swift and Kotlin ones. A hedge
/// that fires at a different moment on one platform means that app has a different latency profile
/// from the one the evaluation describes.
/// </summary>
public class FallbackTranscriberTests
{
    private static Func<CancellationToken, Task<TranscriptionResult>> Backend(
        int delayMs, string text, Exception? failure = null) =>
        async token =>
        {
            await Task.Delay(delayMs, token);
            if (failure is not null) throw failure;
            return new TranscriptionResult(new Transcript(text, "en"), new TokenUsage(), text);
        };

    /// <summary>The common case: the primary answers normally, so the hedge never fires.</summary>
    [Fact]
    public async Task AFastPrimaryIsNeverSecondGuessed()
    {
        var outcome = await new FallbackTranscriber(
            Backend(10, "primary"), "primary", "p-model",
            Backend(10, "secondary"), "secondary", "s-model",
            TimeSpan.FromSeconds(30)).TranscribeAsync();

        Assert.Equal("primary", outcome.Result.Transcript.Text);
        Assert.False(outcome.Attribution.WasFallback);
        Assert.Equal("primary", outcome.Attribution.Provider);
    }

    /// <summary>The case this exists for: the primary stalls, the hedge fires.</summary>
    [Fact]
    public async Task AStalledPrimaryIsOvertakenByTheHedge()
    {
        var outcome = await new FallbackTranscriber(
            Backend(5000, "primary"), "primary", "p-model",
            Backend(20, "secondary"), "secondary", "s-model",
            TimeSpan.FromMilliseconds(20)).TranscribeAsync();

        Assert.Equal("secondary", outcome.Result.Transcript.Text);
        Assert.True(outcome.Attribution.WasFallback, "the caller has to be able to say so");
        Assert.Equal("secondary", outcome.Attribution.Provider);
    }

    /// <summary>
    /// Nothing left to wait for: a failing primary hands over immediately.
    ///
    /// <para>The elapsed time is the claim here, not which transcript came back — "secondary"
    /// arrives either way, just eight seconds later. This case passed against a Swift port that
    /// always slept the full delay, because only the transcript was ever checked. This port has
    /// always hedged on failure; the assertion is what stops it from quietly regressing to
    /// match.</para>
    /// </summary>
    [Fact]
    public async Task AFailingPrimaryFallsBackWithoutWaitingOutTheDelay()
    {
        var clock = Stopwatch.StartNew();
        var outcome = await new FallbackTranscriber(
            Backend(5, "", new ProviderException("boom")), "primary", "p-model",
            Backend(10, "secondary"), "secondary", "s-model",
            TimeSpan.FromSeconds(8)).TranscribeAsync();
        clock.Stop();

        Assert.Equal("secondary", outcome.Result.Transcript.Text);
        Assert.True(outcome.Attribution.WasFallback);
        Assert.True(
            clock.Elapsed < TimeSpan.FromSeconds(1),
            $"the hedge waited out its delay after the primary had already failed ({clock.Elapsed})");
    }

    /// <summary>The primary's error explains the user's configuration, so it is the one shown.</summary>
    [Fact]
    public async Task WhenBothFailThePrimaryErrorSurfaces()
    {
        var error = await Assert.ThrowsAsync<ProviderException>(() =>
            new FallbackTranscriber(
                Backend(5, "", new ProviderException("primary failed")), "primary", "p-model",
                Backend(5, "", new ProviderException("secondary failed")), "secondary", "s-model",
                TimeSpan.FromMilliseconds(1)).TranscribeAsync());

        Assert.Equal("primary failed", error.Message);
    }

    /// <summary>No secondary is the default, and must behave as before this type existed.</summary>
    [Fact]
    public async Task WithoutASecondaryItIsATransparentPassThrough()
    {
        var outcome = await new FallbackTranscriber(
            Backend(5, "primary"), "primary", "p-model").TranscribeAsync();

        Assert.Equal("primary", outcome.Result.Transcript.Text);
        Assert.False(outcome.Attribution.WasFallback);
    }
}
